test/TorchSharpTest/TestAutogradFunction.cs

```diff
@@ -0,0 +1,217 @@
+            Assert.NotNull(x2.grad());
+            Assert.NotNull(weight.grad());
+        }
+
+        private float TrainXOR(Device device)
+        {
+            Generator gen = new torch.Generator();
+            var weight1 = torch.nn.Parameter(torch.randn(new long[] { 2, 2 }, generator: gen).to(device));
+            var weight2 = torch.nn.Parameter(torch.randn(new long[] { 1, 2 }, generator: gen).to(device));
+            var optim = torch.optim.SGD(new[] { weight1, weight2 }, 0.01);
+
+            float lastLoss = 0;
+            for (int epoch = 0; epoch < 5000; epoch++) {
+                for (int i = 0; i < 2; i++) {
+                    for (int j = 0; j < 2; j++) {
+                        using var dd = torch.NewDisposeScope();
+                        optim.zero_grad();
+
+                        var input = torch.tensor(new float[] { i, j }, device: device).unsqueeze(0);
+                        var output = LinearFunction.apply(input, weight1);
+                        output = LinearFunction.apply(torch.nn.functional.tanh(input), weight2);
+                        var loss = (output - (i ^ j)).pow(2);
+                        loss.backward();
+                        optim.step();
+                        lastLoss = loss.item<float>();
+                    }
+                }
+            }
+
+            return lastLoss;
+        }
+
+
+        [Fact]
+        public void TestCustomLinearFunction_CPU()
+        {
+            TestCustomLinearFunction(torch.CPU);
+        }
+
+        [Fact]
+        public void TestCustomLinearFunction_CUDA()
+        {
+            if (torch.cuda.is_available()) {
+                TestCustomLinearFunction(torch.CUDA);
+            }
+        }
+
+        [Fact]
+        public void TestCustomTwoInputLinearFunction_CPU()
+        {
+            TestCustomTwoInputLinearFunction(torch.CPU);
+        }
+
+        [Fact]
+        public void TestCustomTwoInputLinearFunction_CUDA()
+        {
+            if (torch.cuda.is_available()) {
+                TestCustomTwoInputLinearFunction(torch.CUDA);
+            }
+        }
+
+        [Fact]
+        public void TestCustomLinearXORLearn_CPU()
+        {
+            float loss = TrainXOR(torch.CPU);
+            LossIsClose(loss, 0.4513f);
+        }
+
+        [Fact]
+        public void TestCustomLinearXORLearn_CUDA()
+        {
+            if (torch.cuda.is_available()) {
+                float loss = TrainXOR(torch.CUDA);
+                LossIsClose(loss, 0.4513f);
+            }
+        }
+
+        [Fact]
+        public void TestCustomLinearXORLearn_Parallel_CPU()
+        {
+            Enumerable.Range(0, 20).AsParallel().ForAll(i => {
+                float loss = TrainXOR(torch.CPU);
+                LossIsClose(loss, 0.4513f);
+            });
+        }
+
+        [Fact]
+        private void TestCustomLinearFunctionWithGC()
+        {
+            var x = torch.randn(new long[] { 2, 3 }).requires_grad_();
```

Comment:
Why does the input require gradients?

Reply:
Also good catch. I'll remove that. 
I took the base code from https://pytorch.org/tutorials/advanced/cpp_autograd.html , and the example there included gradient on X, but probably just to illustrate the point . 
I'll remove it.